pkg/kubelet/cm/qos_container_manager_linux.go

```diff
@@ -140,9 +143,30 @@ func (m *qosContainerManagerImpl) Start(getNodeAllocatable func() v1.ResourceLis
 		}
 	}, periodicQOSCgroupUpdateInterval, wait.NeverStop)
 
+	go m.worker()
+
 	return nil
 }
 
+func (m *qosContainerManagerImpl) worker() {
+	for {
+		<-m.updateCh
+
+		m.Lock()
```

Comment:
You should not be using both a channel and a lock like this. What behavior are you trying to get?

Reply:
The idea is to buffer all pending update requests in the queue and then reset the queue. This way the queue can grow without blocking the writers. The lock is used to take the current list of pending requests, reset it and prevent writing to the queue from inside UpdateCgroups function. (Using a buffered channel has the downside of being limited by the actual limit that you set when you create the channel, this way we avoid a constant for max channel size in the code that might need to be changed later on or passed via command line).

Reply:
Yeah I figured out what you're trying to do. Please see the singleflight recommendation.

General go code principle: if you find yourself using both a channel and a lock, you can probably find some other design which uses only one or the other and has less contention.

Reply:
I changed the code to use a buffered channel.